### PR TITLE
Fix errors and warnings reported by Clang

### LINKED ISSUE #206: Build fails with Clang
Trying to build current development with Clang 3.8 fails with the following errors:

```
In file included from /home/joseph/src/dash/dash/dash/examples/bench.11.hdf-io/main.cpp:7:
In file included from /home/joseph/src/dash/dash/dash/include/libdash.h:60:
In file included from /home/joseph/src/dash/dash/dash/include/dash/io/HDF5.h:4:
/home/joseph/src/dash/dash/dash/include/dash/io/hdf5/StorageDriver.h:313:19: error: static_assert expression is not an integral constant expression
    static_assert(array.ndim() == pattern_t::ndim(),
```
(`array.ndim()` cannot be used in a static_assert).

Further down the road, Clang has an issue with the Matrix.sub<> function:

```
In file included from /home/joseph/src/dash/dash/dash/test/MatrixTest.cc:1:
In file included from /home/joseph/src/dash/dash/dash/include/libdash.h:36:
In file included from /home/joseph/src/dash/dash/dash/include/dash/Container.h:64:
In file included from /home/joseph/src/dash/dash/dash/include/dash/Matrix.h:600:
/home/joseph/src/dash/dash/dash/include/dash/matrix/internal/Matrix-inl.h:385:15: error: reference to non-static member function must be called
  return _ref.sub<SubDimension>(n);
         ~~~~~^~~
/home/joseph/src/dash/dash/dash/test/MatrixTest.cc:434:26: note: in instantiation of function template specialization 'dash::Matrix<int, 2, long, dash::TilePattern<2,
      dash::MemArrange::ROW_MAJOR, long> >::sub<0>' requested here
    auto column = matrix.sub<0>(col);
                         ^
/home/joseph/src/dash/dash/dash/include/dash/matrix/MatrixRef.h:199:3: note: possible target for call
  sub(size_type n);
  ^
/home/joseph/src/dash/dash/dash/include/dash/matrix/MatrixRef.h:209:3: note: possible target for call
  sub(
```

This seems to be one of the dark corners of the C++ language. The correct code is:

```
  return this->_ref.template sub<1>(offset, extent);
```

Finally, the linking fails at the moment:

```
[ 53%] Linking CXX executable dash-test-mpi
CMakeFiles/dash-test-mpi.dir/test/HDF5MatrixTest.cc.o: In function `std::ostream& dash::operator<< <2, (dash::MemArrange)1, long>(std::ostream&, dash::TilePattern<2, (dash::MemArrange)1, long> const&)':
/home/joseph/src/dash/dash/dash/test/HDF5MatrixTest.cc:(.text._ZN4dashlsILi2ELNS_10MemArrangeE1ElEERSoS2_RKNS_11TilePatternIXT_EXT0_ET1_EE[_ZN4dashlsILi2ELNS_10MemArrangeE1ElEERSoS2_RKNS_11TilePatternIXT_EXT0_ET1_EE]+0x8a): undefined reference to `dash::TilePattern<2, (dash::MemArrange)1, long>::PatternName'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
dash/CMakeFiles/dash-test-mpi.dir/build.make:1232: recipe for target 'dash/dash-test-mpi' failed
```

`PatterName` is a static member, hence 

``` 
pattern.PatternName
``` 
should be 
```
TilePattern<ND,Ar,Index>::PatternName
```

I will create a branch to implement the fixes.

### PR DESCRIPTION
**Pull request by devreal**
Small fixes to make the Clang compiler happy (includes at least one logic error spotted by Clang). Clean build with Clang 3.8.

Fixes #206

### CODE REVIEW

File: dash/test/MatrixTest.cc [hunk 1]
```diff
@@ -414,7 +414,7 @@ TEST_F(MatrixTest, Sub2DimDefault)
   element_t * lit  = matrix.lbegin();
   element_t * lend = matrix.lend();
   LOG_MESSAGE("Local range: lend(%p) - lbegin(%p) = %d",
-              lend, lit, lend - lit);
+              (void*)lend, (void*)lit, lend - lit);
```
Comment (fmoessbauer) [inline dash/test/MatrixTest.cc]: Why don't you use C++ casts?

File: dash/include/dash/GlobRef.h [hunk 2]
```diff
@@ -130,7 +130,7 @@ class GlobRef
 
   inline bool operator!=(const T & value) const
   {
-    return !(*this != value);
+    return !(*this == value);
```
Comment (fmoessbauer) [inline dash/include/dash/GlobRef.h:133]: Good catch!

File: dash/include/dash/io/hdf5/StorageDriver.h [hunk 3]
```diff
@@ -310,7 +310,7 @@ class StoreHDF {
     hdf5_options                                      foptions
                                                         = _get_fdefaults())
   {
-    static_assert(array.ndim() == pattern_t::ndim(),
+    static_assert(ndim == pattern_t::ndim(),
```
Comment (fmoessbauer) [inline dash/include/dash/io/hdf5/StorageDriver.h:313]: It's a bit curious that my implementation worked in gcc.

**Review by fmoessbauer — APPROVED** [review]
In my opinion all your changes are good ones.

File: dash/include/dash/GlobRef.h (hunk 2 above)
Comment (devreal) [reply dash/include/dash/GlobRef.h:133]: Actually, Clang complained about never-ending recursion in this case. Pretty neat.

File: dash/test/MatrixTest.cc (hunk 1 above)
Comment (devreal) [reply dash/test/MatrixTest.cc]: Because they are so verbose :D But I will change it nevertheless.

File: dash/include/dash/GlobRef.h (hunk 2 above)
Comment (fuchsto) [reply dash/include/dash/GlobRef.h:133]: That is darn impressive!

File: dart-impl/mpi/include/dash/dart/mpi/dart_synchronization_priv.h [hunk 4]
```diff
@@ -4,7 +4,7 @@
  * Definition of dart_lock_struct.
  */
 #ifndef DART_ADAPT_SYNCHRONIZATION_PRIV_H_INCLUDED
-#define DART_ADAPT_SYNCHRONZIATION_PRIV_H_INCLUDED
```
Comment (fuchsto) [inline dart-impl/mpi/include/dash/dart/mpi/dart_synchronization_priv.h:7]: Hm, that wasn't me, my vim config wouldn't allow that ... let's see who ...

https://github.com/dash-project/dash/blame/64b7aaf5e0d288d92229e81646fd04468f1f4cfc/dart-impl/mpi/include/dash/dart/mpi/dart_synchronization_priv.h#L7

@fuerlinger 
"KARL? ... KAAAAAAAAARRRL!"
https://www.youtube.com/watch?v=ajsNJtnUb7c

**Pull request merged**